packages/extras/src/lib/components/Sparkles/Sparkles.svelte

```diff
@@ -0,0 +1,102 @@
+
+	export let count: SparklesProperties['count'] = 100;
+	export let speed: SparklesProperties['speed'] = 1;
+	export let opacity: SparklesProperties['opacity'] = 1;
+	export let color: SparklesProperties['color'] = 100;
+	export let size: SparklesProperties['size']= Math.random();
+	export let scale: SparklesProperties['scale'] = 1;
+	export let noise: SparklesProperties['noise'] = 1;
+
+	const isVector = (v: any): v is Vector2 | Vector3 | Vector4 =>
+		v instanceof Vector2 || v instanceof Vector3 || v instanceof Vector4;
+
+	const isFloat32Array = (def: any): def is Float32Array =>
+		def && (def as Float32Array).constructor === Float32Array;
+
+	const normalizeVector = (v: any): number[] => {
+		if (Array.isArray(v)) return v;
+		else if (isVector(v)) return v.toArray();
+		return [v, v, v] as number[];
+	};
+
+	const expandColor = (v: Color) => [v.r, v.g, v.b];
+
+	const positions = Float32Array.from(
+		Array.from({ length: count }, () => normalizeVector(scale).map(randFloatSpread)).flat()
+	);
+
+	const usePropAsIsOrAsAttribute = <T extends any>(
+		count: number,
+		prop?: T | Float32Array,
+		setDefault?: (v: T) => number
+	) => {
+		if (prop !== undefined) {
+			if (isFloat32Array(prop)) {
+				return prop as Float32Array;
+			} else {
+				if (prop instanceof Color) {
+					const a = Array.from({ length: count * 3 }, () => expandColor(prop)).flat();
+					return Float32Array.from(a);
+				} else if (isVector(prop) || Array.isArray(prop)) {
+					const a = Array.from({ length: count * 3 }, () => normalizeVector(prop)).flat();
+					return Float32Array.from(a);
+				}
+
+				return Float32Array.from({ length: count }, () => prop as unknown as number);
+			}
+		}
+		return Float32Array.from({ length: count }, setDefault!);
+	};
+
+	let opacities = usePropAsIsOrAsAttribute<number>(count, opacity);
+	let sizes = usePropAsIsOrAsAttribute<number>(count, size, Math.random);
+	let speeds = usePropAsIsOrAsAttribute<number>(count, speed);
+	let noises = usePropAsIsOrAsAttribute<typeof noise>(count * 3, noise);
+	let colors = usePropAsIsOrAsAttribute<ColorRepresentation>(
+				color === undefined ? count * 3 : count,
+				!isFloat32Array(color) ? new Color(color) : color,
+				() => 1
+			)
+
+	const material = new ShaderMaterial({
+		uniforms: SparkleShader.uniforms,
+		vertexShader: SparkleShader.vertextShader,
+		fragmentShader: SparkleShader.fragmentShader,
+		transparent: true,
+		depthWrite: false,
+	})
+
+	const geometry = new BufferGeometry()
+
+	geometry.setAttribute('position', new BufferAttribute(positions, 3))
+	geometry.setAttribute('size', new BufferAttribute(sizes, 1))
+	geometry.setAttribute('opacity', new BufferAttribute(opacities, 1))
+	geometry.setAttribute('speed', new BufferAttribute(speeds, 1))
+	geometry.setAttribute('color', new BufferAttribute(colors, 3))
+	geometry.setAttribute('noise', new BufferAttribute(noises, 3))
+
+	useFrame((state) => (material.uniforms.time.value = state.clock.elapsedTime));
+
+</script>
+
+<Three
```

Comment:
In order to be able to position the Sparkles with `<Sparkles position={[10, 10, 10]} >` you can pass them $$restProps here.

Reply:
And adjust the types of the component to extend the types of `<Three type={Points}>`

Reply:
So the idea how component types are structured changes a bit with v5. Instead of placing `SparklesProperties` in types/components you need to place the file "Sparkles.svelte.d.ts" right next to "Sparkles.svelte". It's content looks like this:

```ts
	import type { Events, Props, Slots } from '@threlte/core'
	import { SvelteComponentTyped } from 'svelte'
	import type { Points, Vector3, ColorRepresentation } from 'three'

	export default class Grouped extends SvelteComponentTyped<
		Props<Points> & {
			/** Number of particles (default: 100) */
			count: number
			/** Speed of particles (default: 1) */
			speed: number | Float32Array
			/** Opacity of particles (default: 1) */
			opacity: number | Float32Array
			/** Color of particles (default: 100) */
			color: ColorRepresentation | Float32Array
			/** Color of particles (default: 100) */
			size: number | Float32Array
			/** The space the particles occupy (default: 1) */
			scale: number | [number, number, number] | Vector3
			/** The space the particles occupy (default: 1) */
			noise: number | [number, number, number] | Vector3 | Float32Array

			/** Experimental: try to exclude material and geometry */
			material: never
			geometry: never
		},
		Events<Points>,
		Slots<Points>
	> {}
```

This will provide proper type intellisense in Editors like VS Code. Sorry for the confusion, this is a new mechanic also to me.

Reply:
Edit: I changed `Grouped` to `Sparkles` as the exported class name, sorry!